File: bundle-src/src/fragments.ts
// The fragment registry conventions (editor side). A fragment is a static
// piece of markup — typically cut verbatim from a design mockup — that
// provider add-ons register into the shared `@plone/registry` singleton
// from their install function. The fragment block enumerates and renders
// these records; nothing about a fragment lives in Plone content.
//
// `getUtilities` returns records without their registration names, so the
// record itself must carry its `id` (kept equal to the registration name).

export const FRAGMENT_UTILITY_TYPE = 'collective.fragmentsblock.fragment';

export type FragmentRecord = {
  /** Registration name; the value stored on the block node. */
  id: string;
  /** Label shown in the block's fragment picker. */
  title: string;
  /** The fragment markup, injected as-is apart from `${var}` tokens. */
  html: string;
  description?: string;
};

type Registry = {
  registerUtility: (options: {
    type: string;
    name: string;
    method: unknown;
  }) => void;
  getUtility: (options: { type: string; name: string }) => { method?: unknown };
  getUtilities: (options: { type: string }) => Array<{ method?: unknown }>;
};

export function registerFragment(config: Registry, record: FragmentRecord) {
  if (!record?.id || !record.title || typeof record.html !== 'string') {
    throw new Error(
      'registerFragment: a fragment record needs id, title and html',
    );
  }
  config.registerUtility({
    type: FRAGMENT_UTILITY_TYPE,
    name: record.id,
    method: record,
  });
}

function asRecord(utility: { method?: unknown } | undefined) {
  const record = utility?.method as FragmentRecord | undefined;
  return record && record.id && typeof record.html === 'string'
    ? record
    : null;
}

export function getFragment(
  config: Registry,
  id: unknown,
): FragmentRecord | null {
  if (typeof id !== 'string' || !id) return null;
  return asRecord(config.getUtility({ type: FRAGMENT_UTILITY_TYPE, name: id }));
}

export function listFragments(config: Registry): FragmentRecord[] {
  return config
    .getUtilities({ type: FRAGMENT_UTILITY_TYPE })
    .map(asRecord)
    .filter((record): record is FragmentRecord => record !== null)
    .sort((a, b) => a.title.localeCompare(b.title));
}

// ${var} substitution — the "template" half. Semantics are mirrored 1:1 by
// the server renderer (fragments.py substitute()): values come from the
// block's persisted `variables` mapping (so both surfaces see identical
// input), missing variables become empty strings, every value is
// HTML-escaped.
const TOKEN = /\$\{([A-Za-z_][A-Za-z0-9_]*)\}/g;

function escapeHtml(value: unknown): string {
  return String(value)
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#x27;');
}

export function renderFragmentHtml(
  record: FragmentRecord,
  variables?: Record<string, unknown> | null,
): string {
  const values = variables && typeof variables === 'object' ? variables : {};
  return record.html.replace(TOKEN, (_match, name: string) =>
    values[name] == null ? '' : escapeHtml(values[name]),
  );
}

File: src/collective/fragmentsblock/fragments.py
"""Resolve and render registered fragment markup (server side).

A fragment is a static HTML file — typically cut verbatim from a design
mockup — shipped by a provider add-on. The editor half registers id, title
and the same markup into ``@plone/registry`` (see ``bundle-src``); this
module is the classic-rendering counterpart: named ``IFragmentsProvider``
utilities map fragment ids to the raw HTML, and ``substitute`` mirrors the
JS ``renderFragmentHtml`` semantics 1:1 so both surfaces emit identical
markup — ``${var}`` tokens filled from the block's persisted ``variables``
mapping, missing variables as empty strings, every value HTML-escaped.

Everything is fail-soft: an unknown id, an unregistered provider or an
unreadable file degrades to ``None`` (the caller emits an invisible
placeholder), never a broken page.
"""

import html
import math
import re
from logging import getLogger
from pathlib import Path

from zope.component import getUtilitiesFor
from zope.interface import implementer

from collective.fragmentsblock.interfaces import IFragmentsProvider


logger = getLogger(__name__)

# Registration names are filesystem-safe slugs; anything else (e.g. a path
# traversal attempt through the block data) simply does not resolve.
_ID_RE = re.compile(r"^[A-Za-z0-9][A-Za-z0-9_-]*$")

_TOKEN_RE = re.compile(r"\$\{([A-Za-z_][A-Za-z0-9_]*)\}")


@implementer(IFragmentsProvider)
class FragmentsFolder:
    """The stock provider: one ``<fragment_id>.html`` file per fragment.

    Point it at the directory of verbatim mockup files inside the provider
    package — the same files whose ``?raw`` imports the provider's editor
    bundle registers — and register it as a named utility::

        <utility
            name="my.theme"
            provides="collective.fragmentsblock.interfaces.IFragmentsProvider"
            component="my.theme.fragments.provider"
            />
    """

    def __init__(self, directory):
        self.directory = Path(directory)

    def get(self, fragment_id):
        path = self.directory / f"{fragment_id}.html"
        try:
            return path.read_text(encoding="utf-8")
        except OSError:
            return None


def resolve(fragment_id):
    """The raw HTML registered for ``fragment_id``, or ``None``.

    Ids share one flat, site-wide namespace across every provider:
    providers are asked in utility-name order and the first hit wins, so a
    second provider reusing an id shadows the first (the editor-side
    registry, keyed by name, resolves such a clash the same way).
    """
    if not isinstance(fragment_id, str) or not _ID_RE.match(fragment_id):
        return None
    for _name, provider in sorted(getUtilitiesFor(IFragmentsProvider)):
        try:
            markup = provider.get(fragment_id)
        except Exception:
            logger.exception("Fragments provider %r failed for %r", _name, fragment_id)
            continue
        if markup is not None:
            return markup
    return None


def coerce(value):
    """The coercion table both halves implement (``fragments.ts coerce``).

    Values arrive as JSON, so booleans and numbers are ordinary; each one
    is rendered the way JavaScript renders it, because the editor cannot
    render it any other way and the two surfaces must agree. Anything
    outside the table (lists, dicts) renders empty rather than as a
    Python-flavoured string.
    """
    if value is None:
        return ""
    if isinstance(value, str):
        return value
    # before int: bool is an int subclass
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, int):
        return str(value)
    if isinstance(value, float):
        if not math.isfinite(value):
            return ""
        # JS has one number type: 1.0 prints as "1"
        return str(int(value)) if value.is_integer() else repr(value)
    return ""


def substitute(markup, variables):
    """Fill ``${var}`` tokens; mirrors the JS ``renderFragmentHtml``."""
    values = variables if isinstance(variables, dict) else {}

    def _value(match):
        name = match.group(1)
        if name not in values:
            return ""
        # quote=True gives exactly the JS escape set: & < > " ' -> &#x27;
        return html.escape(coerce(values[name]), quote=True)

    return _TOKEN_RE.sub(_value, markup)


def fragment_html(data):
    """The rendered markup of a fragment block dict, or ``None``."""
    markup = resolve((data or {}).get("fragment"))
    if markup is None:
        return None
    return substitute(markup, (data or {}).get("variables"))
